From 2a3a4741916b3775ec8383d9692896fb78aef64d Mon Sep 17 00:00:00 2001
From: Ariel Aviad <aaviad@habana.ai>
Date: Sun, 16 Mar 2025 10:07:56 +0200
Subject: [PATCH] [SW-221257] g3: automate create littlefs flash image

Add python script to generate LittleFS flash
image to load on flash, that represent serdes configuration

Add A JSON for the script the contains the configuration
for the LittleFS image

Change-Id: I009e1c03472c76b13a5d89a474bbf99d5dbaa9d4
Signed-off-by: Ariel Aviad <aaviad@habana.ai>
---

diff --git a/scripts/serdes/gaudi3_serdes_config_flash_memory.json b/scripts/serdes/gaudi3_serdes_config_flash_memory.json
new file mode 100644
index 0000000..8533c94
--- /dev/null
+++ b/scripts/serdes/gaudi3_serdes_config_flash_memory.json
@@ -0,0 +1,432 @@
+{
+	"serdes_table": [
+		{
+			"type": 6,
+			"link_id": "0b00001",
+			"tx_lane_swap" : ["0x0", "0x0", "0x0", "0x0", "0x0", "0x0",
+			"0x0", "0x0", "0x0", "0x0", "0x0", "0x0"],
+			"configurations": [
+				{
+					"index": 0,
+					"tx_pol": 0,
+					"rx_pol": 0,
+					"enable_mask": "0xFFFFFF",
+					"external_mask": "0x0",
+					"autoneg_mask": "0x0"
+				},
+				{
+					"index": 1,
+					"tx_pol": 0,
+					"rx_pol": 0,
+					"enable_mask": "0xFFFFFF",
+					"external_mask": "0x0",
+					"autoneg_mask": "0x0"
+				},
+				{
+					"index": 2,
+					"tx_pol": 0,
+					"rx_pol": 0,
+					"enable_mask": "0xFFFFFF",
+					"external_mask": "0x0",
+					"autoneg_mask": "0x0"
+				},
+				{
+					"index": 3,
+					"tx_pol": 0,
+					"rx_pol": 0,
+					"enable_mask": "0xFFFFFF",
+					"external_mask": "0x0",
+					"autoneg_mask": "0x0"
+				},
+				{
+					"index": 4,
+					"tx_pol": 0,
+					"rx_pol": 0,
+					"enable_mask": "0xFFFFFF",
+					"external_mask": "0x0",
+					"autoneg_mask": "0x0"
+				},
+				{
+					"index": 5,
+					"tx_pol": 0,
+					"rx_pol": 0,
+					"enable_mask": "0xFFFFFF",
+					"external_mask": "0x0",
+					"autoneg_mask": "0x0"
+				},
+				{
+					"index": 6,
+					"tx_pol": 0,
+					"rx_pol": 0,
+					"enable_mask": "0xFFFFFF",
+					"external_mask": "0x0",
+					"autoneg_mask": "0x0"
+				},
+				{
+					"index": 7,
+					"tx_pol": 0,
+					"rx_pol": 0,
+					"enable_mask": "0xFFFFFF",
+					"external_mask": "0x0",
+					"autoneg_mask": "0x0"
+				}
+			]
+		},
+		{
+			"type": 6,
+			"link_id": "0b00100",
+			"tx_lane_swap" : ["0x0", "0x0", "0x0", "0x0", "0x0", "0x0",
+			"0x0", "0x0", "0x0", "0x0", "0x0", "0x0"],
+			"configurations": [
+				{
+					"index": 0,
+					"tx_pol": "0x000000000000",
+					"rx_pol": "0x00000030F010",
+					"enable_mask": "0xFFFFFF",
+					"external_mask": "0x0",
+					"autoneg_mask": "0x3FFF"
+				},
+				{
+					"index": 1,
+					"tx_pol": "0x000000000000",
+					"rx_pol": "0x0C0000000003",
+					"enable_mask": "0xFFFFFF",
+					"external_mask": "0x0",
+					"autoneg_mask": "0xFFF003"
+				},
+				{
+					"index": 2,
+					"tx_pol": "0x000000000000",
+					"rx_pol": "0x0CF20000000F",
+					"enable_mask": "0xFFFFFF",
+					"external_mask": "0x0",
+					"autoneg_mask": "0xFFF003"
+				},
+				{
+					"index": 3,
+					"tx_pol": "0x0100E0000000",
+					"rx_pol": "0x000000000300",
+					"enable_mask": "0xFFFFFF",
+					"external_mask": "0x0",
+					"autoneg_mask": "0x3FFF"
+				},
+				{
+					"index": 4,
+					"tx_pol": "0x000000000000",
+					"rx_pol": "0x000000001060",
+					"enable_mask": "0xFFFFFF",
+					"external_mask": "0x0",
+					"autoneg_mask": "0xC00FFF"
+				},
+				{
+					"index": 5,
+					"tx_pol": "0x000000008000",
+					"rx_pol": "0x00F001F00000",
+					"enable_mask": "0xFFFFFF",
+					"external_mask": "0x0",
+					"autoneg_mask": "0xFFFC00"
+				},
+				{
+					"index": 6,
+					"tx_pol": "0x00000000D000",
+					"rx_pol": "0x030000F00000",
+					"enable_mask": "0xFFFFFF",
+					"external_mask": "0x0",
+					"autoneg_mask": "0xFFFC00"
+				},
+				{
+					"index": 7,
+					"tx_pol": "0x000000000000",
+					"rx_pol": "0x000000000000",
+					"enable_mask": "0xFFFFFF",
+					"external_mask": "0x0",
+					"autoneg_mask": "0xC00FFF"
+				}
+			]
+		},
+		{
+			"type": 7,
+			"link_id": "0b01000",
+			"tx_lane_swap" : ["0x0", "0x0", "0x0", "0x0", "0x0", "0x0", "0x0", "0x0", "0x0", "0x0", "0x0", "0x0"],
+			"configurations": [
+				{
+					"index": 0,
+					"tx_pol": "0x000000000000",
+					"rx_pol": "0x00000030F010",
+					"enable_mask": "0xFFFFFF",
+					"external_mask": "0xFFC000",
+					"autoneg_mask": "0x3FFF"
+				},
+				{
+					"index": 1,
+					"tx_pol": "0x000000000000",
+					"rx_pol": "0x0C0000000003",
+					"enable_mask": "0xFFFFFF",
+					"external_mask": "0x000FFC",
+					"autoneg_mask": "0xFFF003"
+				},
+				{
+					"index": 2,
+					"tx_pol": "0x000000000000",
+					"rx_pol": "0x0CF20000000F",
+					"enable_mask": "0xFFFFFF",
+					"external_mask": "0x000FFC",
+					"autoneg_mask": "0xFFF003"
+				},
+				{
+					"index": 3,
+					"tx_pol": "0x0100E0000000",
+					"rx_pol": "0x000000000300",
+					"enable_mask": "0xFFFFFF",
+					"external_mask": "0xFFC000",
+					"autoneg_mask": "0x3FFF"
+				},
+				{
+					"index": 4,
+					"tx_pol": "0x000000000000",
+					"rx_pol": "0x000000001060",
+					"enable_mask": "0xFFFFFF",
+					"external_mask": "0x3FF000",
+					"autoneg_mask": "0xC00FFF"
+				},
+				{
+					"index": 5,
+					"tx_pol": "0x000000008000",
+					"rx_pol": "0x00F001F00000",
+					"enable_mask": "0xFFFFFF",
+					"external_mask": "0x0003FF",
+					"autoneg_mask": "0xFFFC00"
+				},
+				{
+					"index": 6,
+					"tx_pol": "0x00000000D000",
+					"rx_pol": "0x030000F00000",
+					"enable_mask": "0xFFFFFF",
+					"external_mask": "0x0003FF",
+					"autoneg_mask": "0xFFFC00"
+				},
+				{
+					"index": 7,
+					"tx_pol": "0x000000000000",
+					"rx_pol": "0x000000000000",
+					"enable_mask": "0xFFFFFF",
+					"external_mask": "0x3FF000",
+					"autoneg_mask": "0xC00FFF"
+				}
+			]
+		},
+		{
+			"type": 8,
+			"link_id": "0b11110",
+			"tx_lane_swap": ["0x0", "0x0", "0x0", "0x0", "0x0", "0x0", "0x0", "0x0", "0x0", "0x0", "0x0", "0x0"],
+			"configurations": [
+				{
+					"index": 0,
+					"tx_pol": "0x000000000000",
+					"rx_pol": "0x00000030F010",
+					"enable_mask": "0xFFFFFF",
+					"external_mask": "0x320000",
+					"autoneg_mask": "0x3FFF"
+				},
+				{
+					"index": 1,
+					"tx_pol": "0x000000000000",
+					"rx_pol": "0x0C0000000003",
+					"enable_mask": "0xFFFFFF",
+					"external_mask": "0x000320",
+					"autoneg_mask": "0xFFF003"
+				},
+				{
+					"index": 2,
+					"tx_pol": "0x000000000000",
+					"rx_pol": "0x0CF20000000F",
+					"enable_mask": "0xFFFFFF",
+					"external_mask": "0x000320",
+					"autoneg_mask": "0xFFF003"
+				},
+				{
+					"index": 3,
+					"tx_pol": "0x0100E0000000",
+					"rx_pol": "0x000000000300",
+					"enable_mask": "0xFFFFFF",
+					"external_mask": "0x320000",
+					"autoneg_mask": "0x3FFF"
+				},
+				{
+					"index": 4,
+					"tx_pol": "0x000000000000",
+					"rx_pol": "0x000000001060",
+					"enable_mask": "0xFFFFFF",
+					"external_mask": "0x08C000",
+					"autoneg_mask": "0xC00FFF"
+				},
+				{
+					"index": 5,
+					"tx_pol": "0x000000008000",
+					"rx_pol": "0x00F001F00000",
+					"enable_mask": "0xFFFFFF",
+					"external_mask": "0x00008C",
+					"autoneg_mask": "0xFFFC00"
+				},
+				{
+					"index": 6,
+					"tx_pol": "0x00000000D000",
+					"rx_pol": "0x030000F00000",
+					"enable_mask": "0xFFFFFF",
+					"external_mask": "0x00008C",
+					"autoneg_mask": "0xFFFC00"
+				},
+				{
+					"index": 7,
+					"tx_pol": "0x000000000000",
+					"rx_pol": "0x000000000000",
+					"enable_mask": "0xFFFFFF",
+					"external_mask": "0x08C000",
+					"autoneg_mask": "0xC00FFF"
+				}
+			]
+		},
+		{
+			"type": 9,
+			"link_id": "0b10000",
+			"tx_lane_swap": ["0x0", "0x0", "0x0", "0x0", "0x0", "0x0", "0x0", "0x0", "0x0", "0x0", "0x0", "0x0"],
+			"configurations": [
+				{
+					"index": 0,
+					"tx_pol": "0x000000000000",
+					"rx_pol": "0x00000030F010",
+					"enable_mask": "0xFFFFFF",
+					"external_mask": "0x374000",
+					"autoneg_mask": "0x3FFF"
+				},
+				{
+					"index": 1,
+					"tx_pol": "0x000000000000",
+					"rx_pol": "0x0C0000000003",
+					"enable_mask": "0xFFFFFF",
+					"external_mask": "0x000374",
+					"autoneg_mask": "0xFFF003"
+				},
+				{
+					"index": 2,
+					"tx_pol": "0x000000000000",
+					"rx_pol": "0x0CF20000000F",
+					"enable_mask": "0xFFFFFF",
+					"external_mask": "0x000374",
+					"autoneg_mask": "0xFFF003"
+				},
+				{
+					"index": 3,
+					"tx_pol": "0x0100E0000000",
+					"rx_pol": "0x000000000300",
+					"enable_mask": "0xFFFFFF",
+					"external_mask": "0x374000",
+					"autoneg_mask": "0x3FFF"
+				},
+				{
+					"index": 4,
+					"tx_pol": "0x000000000000",
+					"rx_pol": "0x000000001060",
+					"enable_mask": "0xFFFFFF",
+					"external_mask": "0x1DC000",
+					"autoneg_mask": "0xC00FFF"
+				},
+				{
+					"index": 5,
+					"tx_pol": "0x000000008000",
+					"rx_pol": "0x00F001F00000",
+					"enable_mask": "0xFFFFFF",
+					"external_mask": "0x0001DC",
+					"autoneg_mask": "0xFFFC00"
+				},
+				{
+					"index": 6,
+					"tx_pol": "0x00000000D000",
+					"rx_pol": "0x030000F00000",
+					"enable_mask": "0xFFFFFF",
+					"external_mask": "0x0001DC",
+					"autoneg_mask": "0xFFFC00"
+				},
+				{
+					"index": 7,
+					"tx_pol": "0x000000000000",
+					"rx_pol": "0x000000000000",
+					"enable_mask": "0xFFFFFF",
+					"external_mask": "0x1DC000",
+					"autoneg_mask": "0xC00FFF"
+				}
+			]
+		},
+		{
+			"type": 12,
+			"link_id": "0b11111",
+			"tx_lane_swap": ["0x0", "0x0", "0x0", "0x0", "0x0", "0x0", "0x0", "0x0", "0x0", "0x0", "0x0", "0x0"],
+			"configurations": [
+				{
+					"index": 0,
+					"tx_pol": "0x0000F0001030",
+					"rx_pol": "0x000000000000",
+					"enable_mask": "0XFFFFFFFF",
+					"external_mask": "0x320000",
+					"autoneg_mask": "0xCDFFFF"
+				},
+				{
+					"index": 1,
+					"tx_pol": "0x00002000030F",
+					"rx_pol": "0x000000000000",
+					"enable_mask": "0XFFFFFFFF",
+					"external_mask": "0x000320",
+					"autoneg_mask": "0xFFFCDF"
+				},
+				{
+					"index": 2,
+					"tx_pol": "0x06000000011F",
+					"rx_pol": "0x000000000000",
+					"enable_mask": "0XFFFFFFFF",
+					"external_mask": "0x000320",
+					"autoneg_mask": "0xFFFCDF"
+				},
+				{
+					"index": 3,
+					"tx_pol": "0x000200000100",
+					"rx_pol": "0x000000000000",
+					"enable_mask": "0XFFFFFFFF",
+					"external_mask": "0x320000",
+					"autoneg_mask": "0xCDFFFF"
+				},
+				{
+					"index": 4,
+					"tx_pol": "0x0803000000C0",
+					"rx_pol": "0x000000000000",
+					"enable_mask": "0XFFFFFFFF",
+					"external_mask": "0x08C000",
+					"autoneg_mask": "0xF73FFF"
+				},
+				{
+					"index": 5,
+					"tx_pol": "0x013F0F303000",
+					"rx_pol": "0x000000000000",
+					"enable_mask": "0XFFFFFFFF",
+					"external_mask": "0x00008C",
+					"autoneg_mask": "0xFFFF73"
+				},
+				{
+					"index": 6,
+					"tx_pol": "0x0F80C0F01500",
+					"rx_pol": "0x000000000000",
+					"enable_mask": "0XFFFFFFFF",
+					"external_mask": "0x00008C",
+					"autoneg_mask": "0xFFFF73"
+				},
+				{
+					"index": 7,
+					"tx_pol": "0x0C0104000003",
+					"rx_pol": "0x000000000000",
+					"enable_mask": "0XFFFFFFFF",
+					"external_mask": "0x08C000",
+					"autoneg_mask": "0xF73FFF"
+				}
+			]
+		}
+	]
+}
\ No newline at end of file
diff --git a/scripts/serdes/serdes_generate_littlefs_flash_memory.py b/scripts/serdes/serdes_generate_littlefs_flash_memory.py
new file mode 100755
index 0000000..b9fea9f
--- /dev/null
+++ b/scripts/serdes/serdes_generate_littlefs_flash_memory.py
@@ -0,0 +1,87 @@
+#!/usr/bin/python3
+
+import json
+import sys
+from littlefs import LittleFS
+
+JSON_FILE_DEFAULT = 'gaudi3_serdes_config_flash_memory.json'
+OUTPUT_FILE_DEFAULT = 'FlashMemory.bin'
+
+def json_field_as_int(json, field_name):
+    val = json[field_name]
+
+    # if the value is a string, convert it to an integer in base 16
+    if isinstance(val, str):
+        val = int(val, 16)
+
+    return val
+
+def parse_config_json(file_path, output_file, debug=False):
+    with open(file_path, 'r') as file:
+        data = json.load(file)
+
+    fs = LittleFS(block_size=4096, block_count=16)
+    for serdes_entry in data['serdes_table']:
+        type_value = serdes_entry['type']
+        link_id_value = int(serdes_entry['link_id'], 2)
+        tx_lane_swap = [int(lane, 16) for lane in serdes_entry['tx_lane_swap']]
+
+        for config in serdes_entry['configurations']:
+            index = config['index']
+            tx_pol = json_field_as_int(config, 'tx_pol')
+            rx_pol = json_field_as_int(config, 'rx_pol')
+            enable_mask = json_field_as_int(config, 'enable_mask')
+            external_mask = json_field_as_int(config, 'external_mask')
+            autoneg_mask = json_field_as_int(config, 'autoneg_mask')
+
+            if debug:
+                print(f"Index: {index}")
+                print(f"Type: {type_value}")
+                print(f"Link ID: {link_id_value}")
+                print(f"TX Pol: {hex(tx_pol)}")
+                print(f"RX Pol: {hex(rx_pol)}")
+                print(f"Enable Mask: {hex(enable_mask)}")
+                print(f"External Mask: {hex(external_mask)}")
+                print(f"Autoneg Mask: {hex(autoneg_mask)}")
+                print(f"TX Lane Swap: {tx_lane_swap}")
+                print()
+
+            output_dir = '.'
+            file_name = f"{output_dir}/serdes-{link_id_value}-{index}"
+            with fs.open(file_name, 'wb') as fh:
+                fh.write(type_value.to_bytes(2, byteorder='little'))
+                fh.write(tx_pol.to_bytes(8, byteorder='little'))
+                fh.write(rx_pol.to_bytes(8, byteorder='little'))
+                fh.write(enable_mask.to_bytes(4, byteorder='little'))
+                fh.write(external_mask.to_bytes(4, byteorder='little'))
+                fh.write(autoneg_mask.to_bytes(4, byteorder='little'))
+                for lane in tx_lane_swap:
+                    fh.write(lane.to_bytes(2, byteorder='little'))
+
+    with open(output_file, 'wb') as fh:
+        fh.write(fs.context.buffer)
+
+def print_help():
+    print("Usage: parse_json_no_enums_names.py [<path_to_json_file>] [<output_file>] [--debug]")
+    print()
+    print("Arguments:")
+    print(f"  <path_to_json_file>  Optional path to the input JSON file containing Serdes configuration. Default is {JSON_FILE_DEFAULT}.")
+    print(f"  <output_file>        Optional path to the output file where the LittleFS image will be saved. Default is {OUTPUT_FILE_DEFAULT}.")
+    print("  --debug              Optional flag to enable debug mode for detailed output.")
+    print("  --help               Prints this help message.")
+
+if __name__ == "__main__":
+    if '--help' in sys.argv:
+        print_help()
+        sys.exit(0)
+
+    json_file_path = sys.argv[1] if len(sys.argv) > 1 and not sys.argv[1].startswith('--') else JSON_FILE_DEFAULT
+    output_file_path = sys.argv[2] if len(sys.argv) > 2 and not sys.argv[2].startswith('--') else OUTPUT_FILE_DEFAULT
+    debug_mode = '--debug' in sys.argv
+
+    print(f"Reading JSON file: {json_file_path}")
+    print(f"Output file: {output_file_path}")
+    if debug_mode:
+        print("Debug mode enabled")
+
+    parse_config_json(json_file_path, output_file_path, debug_mode)
\ No newline at end of file
